lib/aws/rails/middleware/elastic_beanstalk_sqsd.rb

```diff
@@ -28,25 +30,64 @@ def call(env)
           periodic_task?(request) ? execute_periodic_task(request) : execute_job(request)
         end
 
+        def shutdown(timeout = nil)
+          return unless @executor
+
+          @logger.info("Shutting down SQS EBS background job executor. Timeout: #{timeout}")
+          @executor.shutdown
+          clean_shutdown = @executor.wait_for_termination(timeout)
+          @logger.info("SQS EBS background executor shutdown complete. Clean: #{clean_shutdown}")
+        end
+
         private
 
+        def init_executor
+          threads = Integer(ENV.fetch('AWS_PROCESS_BEANSTALK_WORKER_THREADS',
+                                      Concurrent.available_processor_count || Concurrent.processor_count))
+          puts "Running with threads: #{threads}"
+          options = {
+            max_threads: threads,
+            max_queue: 1,
+            fallback_policy: :abort # Concurrent::RejectedExecutionError must be handled
+          }
+          @executor = Concurrent::ThreadPoolExecutor.new(options)
+          at_exit { shutdown }
```

Comment:
Is this the right place for this (on initialize)? But also, looking at the ThreadPoolExecutor docs, it looks like auto_terminate is true and will register an at_exit handler already. I'm assuming you only did this for logging?

Reply:
this behavior seems to have changed in more recent versions of concurrent-ruby and without setting our own at_exit, we don't get clean shutdown.  I've kept this here because I'm not sure if theres a better place for it...